File: 2024/day/11/main.py
from functools import lru_cache

def read_input(file):
  with open(file, 'r') as f:
    first_line = f.readline().strip()
    return [int(x) for x in first_line.split(' ')]

# This single line trivializes the problem completely
@lru_cache(maxsize=None)
def count_stones(stone, blinks):
  if blinks == 0:
    return 1

  if stone == 0:
    return count_stones(1, blinks - 1)

  str_stone = str(stone)
  if len(str_stone) % 2 == 0:
    left = int(str_stone[:len(str_stone) // 2])
    right = int(str_stone[len(str_stone) // 2:])
    return count_stones(left, blinks - 1) + count_stones(right, blinks - 1)
  
  return count_stones(stone * 2024, blinks - 1)

if __name__ == '__main__':
  stones = read_input('input.txt')

  blinks = 25
  stone_count = 0
  for stone in stones:
    stone_count += count_stones(stone, blinks)

  print("(Part one) stone count: " + str(stone_count))

  blinks = 75
  stone_count = 0
  for stone in stones:
    stone_count += count_stones(stone, blinks)

  print("(Part two) stone count: " + str(stone_count))

File: 2024/day/11/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	stones, err := readInput()
	if err != nil {
		fmt.Println(err)
		return
	}

	blinks := 25
	sum := 0
	for _, stone := range stones {
		sum += countStones(stone, blinks)
	}
	fmt.Printf("(Part one) Stone count: %d\n", sum)

	blinks = 75
	sum = 0
	for _, stone := range stones {
		sum += countStones(stone, blinks)
	}
	fmt.Printf("(Part two) Stone count: %d\n", sum)
}

func readInput() ([]int, error) {
	file, err := os.ReadFile("input.txt")
	if err != nil {
		return nil, err
	}
	numbers := strings.Split(string(file), " ")
	values := make([]int, len(numbers))
	for i, n := range numbers {
		value, err := strconv.Atoi(n)
		if err != nil {
			return nil, err
		}
		values[i] = value
	}

	return values, nil
}

var cache = map[string]int{}

func countStones(stone, blinks int) int {
	key := fmt.Sprintf("s%d-b%d", stone, blinks)

	if hit, ok := cache[key]; ok {
		return hit
	}

	if blinks < 0 {
		return 0
	}

	if blinks == 0 {
		return 1
	}

	if stone == 0 {
		result := countStones(1, blinks-1)
		cache[key] = result
		return result
	}

	stoneAsString := strconv.Itoa(stone)
	if len(stoneAsString)%2 == 0 {
		left := stoneAsString[:len(stoneAsString)/2]
		right := stoneAsString[len(stoneAsString)/2:]
		leftAsInt, err := strconv.Atoi(left)
		if err != nil {
			panic(err)
		}
		rightAsInt, err := strconv.Atoi(right)
		if err != nil {
			panic(err)
		}
		result := countStones(leftAsInt, blinks-1) + countStones(rightAsInt, blinks-1)
		cache[key] = result
		return result
	}

	result := countStones(stone*2024, blinks-1)
	cache[key] = result
	return result
}
